### PR TITLE
feat(*) merge Helm and Deployment Manager

### PR DESCRIPTION
**Pull request by adamreese**
The scope of this PR is to merge the helm and deployment-manager code repository.  I have updated some of the tooling for builds and testing.  We can open new issues/PR's for the remaining items.

Closes: https://github.com/kubernetes/deployment-manager/issues/286

### CODE REVIEW

**Comment by adamreese** [comment]
Import paths are for existing repository name.

**Comment by jackgr** [comment]
Looking...

**Comment by technosophos** [comment]
The CLA bot seems to have gotten a few of @michelleN and my commits wrong. That could have been part of the engineyard.com -> deis.com email change we had to make.

File: pkg/dm/install.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	out, err := runner.GetByKind("rc", "manager-rc", "dm")
+	if err != nil {
+		format.Err("Installation not found: %s %s", out, err)
+		return false
+	}
+	return true
+}
+
+// InstallYAML is the installation YAML for DM.
+const InstallYAML = `
```
Comment (jackgr) [inline pkg/dm/install.go]: This content should be externalized as a file, not embedded in this binary.

File: rootfs/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+```
+rootfs/bin -> /bin
+rootfs/usr/local/share -> /usr/local/share
+```
+
+## Dockerfile
+
+A Dockerfile in the rootfs is used to build the image. Where possible,
+compilation should not be done in this Dockerfile, since we are
+interested in deploying the smallest possible images.
```
Comment (jackgr) [inline rootfs/README.md]: We should build in a container for reproducible builds. To keep the size of the deployed container small, we can copy the binary from the build container to the deployed container.

File: pkg/kubectl/kubectl.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package kubectl
```
Comment (jackgr) [inline pkg/kubectl/kubectl.go]: Nice improvement of kubectl interface.

File: pkg/chart/testdata/ill/templates/wordpress.jinja [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #helm:generate dm_template
 {% set PROPERTIES = properties or {} %}
-{% set PROJECT = PROPERTIES['project'] or 'dm-k8s-prod' %}
+{% set PROJECT = PROPERTIES['project'] or 'dm-k8s-testing' %}
```
Comment (jackgr) [inline pkg/chart/testdata/ill/templates/wordpress.jinja]: This change and the others like it do not seem to affect any other content. Why do we have to revert them?

File: get-install.sh [hunk 5]
```diff
@@ -15,13 +15,11 @@
...
-DEFAULT_TAG=v1.2
-DEFAULT_BINARY=${GOPATH}/bin/dm
-DEFAULT_PLATFORM=$(uname | tr '[:upper:]' '[:lower:]')
-DEFAULT_ARCH=$(uname -m)
+PLATFORM=$(uname | tr '[:upper:]' '[:lower:]')
+ARCH=$(uname -m)
 
 STORAGE_URL=http://get-dm.storage.googleapis.com
-ZIP=dm-${TAG:-DEFAULT_TAG}-${PLATFORM:-DEFAULT_PLATFORM}-${ARCH:-DEFAULT_ARCH}.zip
+ZIP=dm-latest-${PLATFORM}-${ARCH}.zip
```
Comment (jackgr) [inline get-install.sh]: Why do we have to revert these changes?

File: cmd/helm/chart.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
+package main
```
Comment (jackgr) [inline cmd/helm/chart.go]: Would appreciate a walk through of the helm commands.

File: Makefile [hunk 7]
```diff
@@ -12,41 +12,80 @@
...
 
-.PHONY : all build test clean $(TARGETS) $(SUBDIRS_TARGETS) .project .docker
+GO_DIRS ?= $(shell glide nv -x )
+GO_PKGS ?= $(shell glide nv)
 
+.PHONY: build
 build:
-	go get -v $(GO_DEPS)
-	go install -v $(GO_DEPS)
+	@scripts/build-go.sh
```
Comment (jackgr) [inline Makefile]: This is a substantial change to the build process. Would appreciate a walk through.

File: cmd/dm/dm.go [hunk 8]
```diff
@@ -272,7 +272,7 @@ func callService(path, method, action string, reader io.ReadCloser) {
 	var URL *url.URL
 	URL, err := url.Parse(*service)
 	if err != nil {
-		panic(fmt.Errorf("cannot parse url (%s): %s\n", *service, err))
+		panic(fmt.Errorf("cannot parse url (%s): %s\n", path, err))
```
Comment (vaikas-google) [inline cmd/dm/dm.go]: why this change, isn't it a failure to parse *service in line 273?

File: pkg/dm/install.go (hunk 1 above)
Comment (adamreese) [inline pkg/dm/install.go]: This was added to the requirements so the user could download the binary and install server components with `helm dm install`.

File: cmd/helm/deployment.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+		Usage: "Perform deployment-centered operations.",
+		Subcommands: []cli.Command{
+			{
+				Name:      "config",
+				Usage:     "Dump the configuration file for this deployment.",
+				ArgsUsage: "DEPLOYMENT",
+			},
+			{
+				Name:      "manifest",
+				Usage:     "Dump the Kubernetes manifest file for this deployment.",
```
Comment (vaikas-google) [inline cmd/helm/deployment.go]: there can be multiple manifests, so should this be Dump the latest? or something like that.

File: cmd/helm/helm.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	app := cli.NewApp()
+	app.Name = "helm"
+	app.Version = version
+	app.Usage = `Deploy and manage packages.`
+	app.Commands = commands
+
+	// TODO: make better
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
+			Name:   "host,u",
```
Comment (vaikas-google) [inline cmd/helm/helm.go]: is this a typo? Or some convention that I'm just not aware of?

File: pkg/chart/testdata/ill/templates/wordpress.jinja (hunk 4 above)
Comment (adamreese) [inline pkg/chart/testdata/ill/templates/wordpress.jinja]: git is trying to rebase from initial commit of the helm repository.  It _can_ be rebased, but I would need to resolve the conflicts for 68 commits.  There are some other creative ways to resolve but it would still cost time.

File: cmd/manager/manager/typeresolver_test.go [hunk 11]
```diff
@@ -287,31 +287,31 @@ resources:
...
 			Content: "my-content-2"},
 	}
 
 	downloadResponses := map[string]registry.DownloadResponse{
-		"https://raw.githubusercontent.com/kubernetes/application-dm-templates/master/common/replicatedservice/v1/replicatedservice.py":        registry.DownloadResponse{Err: nil, Code: http.StatusOK, Body: "my-content"},
-		"https://raw.githubusercontent.com/kubernetes/application-dm-templates/master/common/replicatedservice/v1/replicatedservice.py.schema": registry.DownloadResponse{Err: nil, Code: http.StatusNotFound, Body: ""},
-		"https://raw.githubusercontent.com/kubernetes/application-dm-templates/master/common/replicatedservice/v2/replicatedservice.py":        registry.DownloadResponse{Err: nil, Code: http.StatusOK, Body: "my-content-2"},
-		"https://raw.githubusercontent.com/kubernetes/application-dm-templates/master/common/replicatedservice/v2/replicatedservice.py.schema": registry.DownloadResponse{Err: nil, Code: http.StatusNotFound, Body: ""},
+		"https://raw.githubusercontent.com/kubernetes/application-dm-templates/master/common/replicatedservice/v1/replicatedservice.py":        {Err: nil, Code: http.StatusOK, Body: "my-content"},
```
Comment (vaikas-google) [inline cmd/manager/manager/typeresolver_test.go]: Thanks much for cleaning all these up, my bad :)

File: examples/wordpress/README.md [hunk 12]
```diff
@@ -42,7 +42,7 @@ The template contains the following variables:
 
 ```
 {% set PROPERTIES = properties or {} %}
-{% set PROJECT = PROPERTIES['project'] or 'dm-k8s-prod' %}
+{% set PROJECT = PROPERTIES['project'] or 'dm-k8s-testing' %}
```
Comment (vaikas-google) [inline examples/wordpress/README.md]: Just want to make sure these are all intentional, and not left over? I think @jackgr changed all these?

File: cmd/dm/dm.go (hunk 8 above)
Comment (adamreese) [inline cmd/dm/dm.go]: This diff is from reverting commit https://github.com/kubernetes/deployment-manager/commit/6e1821c4bb1c0b927fb99fa5ac1325bf41267db7

File: cmd/manager/manager/typeresolver_test.go (hunk 11 above)
Comment (adamreese) [inline cmd/manager/manager/typeresolver_test.go]: Thanks goes to `gofmt -s`

File: examples/wordpress/README.md (hunk 12 above)
Comment (adamreese) [inline examples/wordpress/README.md]: These changes were made after the code freeze and became a rebasing issue.  There is comment describing the issue in this thread.  :point_up: that way, I think.

Comment (vaikas-google) [inline examples/wordpress/README.md]: :) Very good, sorry for the noise, just wanted to make sure it was intentional.

File: cmd/manager/manager/typeresolver_test.go (hunk 11 above)
Comment (vaikas-google) [inline cmd/manager/manager/typeresolver_test.go]: :) Well, was news to me, so thanks for pointing that out then. So there...

**Comment by jackgr** [comment]
@sparkprime has tracked down the CLA situation. Thanks, @sparkprime! 

The bot is not happy because some of the contributors to this PR (other than the sender, which was @adamreese) have individual CLAs and/or multiple corporate CLAs (i.e., Engine Yard and Deis), which are different from the sender's CLAs.

We need all of the authors of this commit who consent to this PR being submitted to add a comment to this PR stating that consent. When that's done, we will manually set the cla tag to yes.

**Comment by michelleN** [comment]
I hereby give my consent. 
Also, I may be the culprit.

**Comment by sgoings** [comment]
I know I have an individual and corporate CLAs as well. I apologize for my attempt to overachieve. :wink: 

I hereby give my consent to this PR.

**Comment by technosophos** [comment]
I hereby consent.

**Comment by jackgr** [comment]
We heard back from our Open Source Program Office. Thanks, @sparkprime for asking them for guidance. Here's what they said:

> Matt and Seth are both covered by the EngineYard CLA.  Adam and Michelle do not have CLAs at all under the email addresses identified in their commits.  Each has individual CLA which is connected to their GitHub accounts.  Based on the information available in git data, there is no indication that either is an EngineYard employee.  Looking at Michelle's individual CLA, I see that is was signed with her EngineYard email, so when I look her up by email I then see that she is covered by the EngineYard corporate CLA.  I can't find anything indicating that Adam is an EngineYard employee aside from the company listed on his public GitHub profile.

> The problem here is that they didn't use the email address in their git commits that is covered by the EngineYard CLA [so the bot wasn't able to connect the dots].  This is covered under "How do I add authorized contributors for my company?" on https://cla.developers.google.com/about.

> ...[also] there's no way to know whether Adam Reese is even covered by the EngineYard CLA.

> Simply removing [Michelle's] CLA wouldn't actually solve this, since [the bot] wouldn't be able to find her.  Her commits would also need to be updated to use her EngineYard account.  Instructions for doing that can be found at https://help.github.com/articles/changing-author-info/.
